### DISCUSSION TITLE
ROS_PACKAGE_PATH environment variable not read correctly

### DISCUSSION THREAD

**Discussion by rg-gravis** [foxglove/community#1038] [Bugs]
I've been using the `ROS_PACKAGE_PATH` setting in the desktop app to provide a location to read URDFs and meshes from, but now I'm trying to switch to using the environment variable instead of the setting. I can confirm that the variable is set with `echo $ROS_PACKAGE_PATH`, and the `ROS_PACKAGE_PATH` setting shows the value of the environment variable as grayed-out "default" text, but when I try to actually visualize a bag it can't find the meshes. I know the path is correct because it does find the meshes if I just type the exact same path into the setting field.

On Ubuntu 20.04, foxglove studio version 2.12.1

Console output:
```
initializing Sentry in main
Loaded language from settings: en
Set language: en
Foxglove Studio 2.12.1
Could not set app as handler for foxglove://
Elapsed (ms) until new AppWindow: 130
New AppWindow 1
window.loadURL(file:///opt/Foxglove Studio/resources/app.asar/renderer/index.html)
window URL loaded
load-pending-files
injectFilesToOpen: no files - skipping
Renderer initialized in 744ms
Load: package://my_urdf/meshes/BASE.stl
ROS_PACKAGE_PATH 
Error: ROS_PACKAGE_PATH not set
    at /opt/Foxglove Studio/resources/app.asar/main/main.js:260:55039
    at AsyncFunction.<anonymous> (node:electron/js2c/browser_init:2:52242)
```

**Comment by defunctzombie** [comment]
Private tracking issue: https://linear.app/foxglove/issue/FG-8759/ros-package-path-environment-variable-not-read-correctly-1038

**Comment by achim-k** [comment]
I just tested this and as far as I can tell, everything works as expected. See the steps below for how I tested this:

1. Clone a repository with a URDF and some meshes: `git clone https://github.com/Hilti-Research/phasma_description`
2. Open the Foxglove app with `ROS_PACKAGE_PATH` set to the current working directory: `ROS_PACKAGE_PATH=$PWD foxglove-studio`
3. Add a 3D panel and create a custom layer URDF with the following package URL: `package://phasma_description/urdf/phasma.urdf`
4. Observe that the app correctly locates the URDF and all referenced meshes
5. Close the app and reopen it with a different `ROS_PACKAGE_PATH` and observe that the URDF does not load

**Reply by rg-gravis** [reply]
These steps do not produce a working custom URDF layer for me:
```bash
❯ ROS_PACKAGE_PATH=$PWD foxglove-studio
initializing Sentry in main
Loaded language from settings: en
Set language: en
Foxglove Studio 2.13.0
Could not set app as handler for foxglove://
Elapsed (ms) until new AppWindow: 164
New AppWindow 1
window.loadURL(file:///opt/Foxglove Studio/resources/app.asar/renderer/index.html)
window URL loaded
load-pending-files
injectFilesToOpen: no files - skipping
Renderer initialized in 1270ms
Loaded language from settings: en
Set language: en
Load: package://phasma_description/urdf/phasma.urdf
ROS_PACKAGE_PATH 
Error: ROS_PACKAGE_PATH not set
    at /opt/Foxglove Studio/resources/app.asar/main/main.js:248:55019
    at AsyncFunction.<anonymous> (node:electron/js2c/browser_init:2:52242)
```
and the URDF does not render, and shows a "Failed to load URDF" error

**Reply by achim-k** [reply]
What OS are you using and how did you install Foxglove (snap, apt, ...)?
Does `echo $PWD` print the current working directory? 
If not, maybe try setting the path manually, e.g. `ROS_PACKAGE_PATH=/foo/bar/bas foxglove-studio`

**Reply by rg-gravis** [reply]
I'm on Ubuntu 20.04, with Foxglove installed via `apt`. `echo $PWD` works, and in fact I see my current directory as the grayed-out default value of ROS_PACKAGE_PATH in the settings: 
![Mon 23 Sep 14:26:22 CEST 2024](https://github.com/user-attachments/assets/6f699311-bea3-464a-be61-599bcb10687b)

**Reply by achim-k** [reply]
That's strange :thinking: 
It looks like the ROS_PACKAGE_PATH is not set for all components of the app. I will investigate this further.

For reference, this is the stdout that I am getting:
```
ROS_PACKAGE_PATH=$PWD foxglove-studio 
initializing Sentry in main
Loaded language from settings: undefined
Set language: en
Foxglove Studio 2.13.0
Could not set app as handler for foxglove://
Elapsed (ms) until new AppWindow: 257
New AppWindow 1
window.loadURL(file:///opt/Foxglove Studio/resources/app.asar/renderer/index.html)
window URL loaded
load-pending-files
injectFilesToOpen: no files - skipping
Renderer initialized in 2335ms
[853704:0923/142045.303318:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 1 times!
[853704:0923/142047.057285:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 2 times!
[853704:0923/142048.186459:ERROR:gl_surface_presentation_helper.cc(260)] GetVSyncParametersIfAvailable() failed for 3 times!
Load: package://phasma_description/urdf/phasma.urdf
ROS_PACKAGE_PATH /home/achim/Desktop
searching /home/achim/Desktop for packages
Resolved: /home/achim/Desktop/phasma_description/urdf/phasma.urdf
```

**Reply by defunctzombie** [reply]
@rg-gravis Have you tried setting the path explicitly like suggested above

>ROS_PACKAGE_PATH=/foo/bar/bas foxglove-studio

And then trying to open the URDF to see if the logs in the terminal show this dummy ROS_PACKAGE_PATH?

We've so far been unable to reproduce. A few things that might help:

- Which shell are you using?
- Can you confirm you are running these commands from your `/home/rye` directory?
- Does manually updating the setting in the app text field work (rather than letting it auto-detect via the env var)?
- Can you try doing `export ROS_PACKAGE_PATH=/foo/bar/bas` and then running `foxglove-studio` as a separate command in the same terminal session/window?

**Reply by rg-gravis** [reply]
To answer your questions:
- Every log I've posted has had a path set, and none of them have the path show up in the terminal logs
- `zsh`
- Yes
- Yes
- No difference

To back up a bit, the actual goal I'm trying to accomplish here is to automate Foxglove Studio setup for my organization using Ansible, and to that end I'm adding the path that contains our URDFs to `~/.profile`, i.e. `export ROS_PACKAGE_PATH="$HOME/git/core/assets/urdfs"`, which should set this variable for all applications, whether launched from the command line or the desktop environment, so the shell used is irrelevant.

https://github.com/user-attachments/assets/78f733aa-c695-44d1-b086-a46208576047

In this video you can see that I start with that path manually entered into the desktop settings, and the phasma URDF renders fine, and then I go to settings and delete the manually entered path, and instead the environment variable, which has the exact same value, appears as grayed-out text. I then go back to the view, and the URDF is not found. I then re-enter the exact same path in the settings, and it works again.

For trying to reproduce:
```
❯ uname -a
Linux rye-ThinkPad-ubuntu 5.15.0-122-generic #132~20.04.1-Ubuntu SMP Fri Aug 30 15:50:07 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
Version: Foxglove Studio 2.14.0 (also occurred on 2.12.1 and 2.13.0), installed via apt
I also have chromium installed on the system via snap if that's relevant.

**Reply by rg-gravis** [reply]
Just replicated the above behavior on a coworker's laptop (same laptop model and OS)

**Reply by defunctzombie** [reply]
Thanks for going through those steps. We will continue our investigation.

One more item to confirm. If you open a new shell and run `env | grep ROS_PACKAGE_PATH` do you see your expected value from the `~/.profile` file?

**Reply by rg-gravis** [reply]
Yes, I do.

**Reply by defunctzombie** [reply]
I've tried more ways to reproduce (using v2.14.0 installed via apt) and so far don't produce the same results you have experienced. I have tried in `bash` as well as `zsh` (with a default `zsh` config).

Some things I have noticed:

In your first post you indicated seeing console logs like the following:
```
Load: package://my_urdf/meshes/BASE.stl
ROS_PACKAGE_PATH 
Error: ROS_PACKAGE_PATH not set
```

When I run the app with no `ROS_PACKAGE_PATH` env var, I see
```
Load: package://my_urdf/meshes/BASE.stl
ROS_PACKAGE_PATH undefined
Error: ROS_PACKAGE_PATH not set
```

Notice the additional `undefined` after the `ROS_PACKAGE_PATH`. This makes me think that part of the app sees the path as an empty string.

---

I added an entry to my `~/.profile` file with `export ROS_PACKAGE_PATH=$HOME/scratch` (did a logout and login cycle) and saw the correct output in app settings _and_ in the console when trying to load a URDF.

I have no explanation for why you are seeing this but we've had multiple people try this on their Linux machines and they could not reproduce the same results you experience.
